cilium: bpf-based hostport implementation

borkmann opened the pull request:
See commit msg.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/10592)
<!-- Reviewable:end -->

borkmann committed:
cilium, bpf: add native HostPort mapping support

In order to use hostPort today, users can only consume it via deploying
Cilium through --set global.cni.chainingMode=portmap in order to then
have iptables set up for each pod that specifies a hostPort. We almost
have all the BPF infrastructure in place for natively supporting BPF
based hostPort. This work implements the hostPort feature as a service
mapping. It is part of the kube-proxy-free implementation since it
depends on NodePort infrastructure.

Example:

  # ./daemon/cilium-agent --identity-allocation-mode=crd --enable-ipv6=true --enable-ipv4=true --disable-envoy-version-check=true --tunnel=disabled --k8s-kubeconfig-path=$HOME/.kube/config --kube-proxy-replacement=strict --enable-l7-proxy=false

  # cat hostport.yaml
  apiVersion: v1
  kind: Pod
  metadata:
    name: nginx
  spec:
    containers:
      - name: nginx
        image: nginx
        ports:
          - containerPort: 80
            hostPort: 90
  # kubectl apply -f ./hostport.yaml
  pod/nginx created

  # ./cilium/cilium service list
  ID   Frontend            Service Type   Backend
  1    10.96.0.1:443       ClusterIP      1 => 192.168.178.29:6443
  2    10.96.0.10:53       ClusterIP      1 => 10.29.6.103:53
                                          2 => 10.29.24.228:53
  3    10.96.0.10:9153     ClusterIP      1 => 10.29.6.103:9153
                                          2 => 10.29.24.228:9153
  4    192.168.178.29:90   HostPort       1 => 10.29.245.35:80

From own node:

  # curl 192.168.178.29:90
  <!DOCTYPE html>
  <html>
  <head>
  [...]

From remote node:

  root@tank:~# curl 192.168.178.29:90
  <!DOCTYPE html>
  <html>
  <head>
  [...]

The HostPort implementation is a hybrid of ClusterIP and NodePort in
terms of implementation in that making it reachable on the own node is
just a single service map entry (as opposed to further exposing it via
loopback address, etc) and for external traffic the backend is always
local to the node and can be mapped from any port.

Fixes: #10359
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

tklauser commented on cilium/cmd/service_update.go:
```diff
@@ -97,12 +106,14 @@ func updateService(cmd *cobra.Command, args []string) {
-	if k8sExternalIPs && k8sNodePort {
-		Fatalf("Cannot set both --k8s-external and --k8s-node-port for a service")
+	if boolToInt(k8sExternalIPs)+boolToInt(k8sNodePort)+boolToInt(k8sHostPort) > 1 {
+		Fatalf("Cannot set more than one of --k8s-external, --k8s-node-port, --k8s-host-port for a service")
```
Nit, maybe rephrase this to:

```suggestion
		Fatalf("Can only set one of --k8s-external, --k8s-node-port, --k8s-host-port for a service")
```

borkmann committed:
cilium, helm, docs: update kube-proxy free's gsg to document HostPort

Document the new HostPort mapping feature under the advanced section and
add Helm support for it. I've also added a small setup validation as I
think it's quite useful.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
tests: add HostPort connectivity tests to k8sT service tests

Add connectivity test cases from internal and external.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

brb commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -321,6 +322,92 @@ then the same range must be passed to Cilium via the ``global.nodePort.range``
+
+.. parsed-literal::
+
+    apiVersion: apps/v1
```
Nit: `kubectl run nginx --generator=run-pod/v1 --image=nginx --replicas=1 --port=80 --hostport=8080`?

borkmann replied:
I was referring to the example yaml from earlier in the same doc, so I kept it as is but with the added `hostPort`.

brb commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -321,6 +322,92 @@ then the same range must be passed to Cilium via the ``global.nodePort.range``
+      selector:
+        matchLabels:
+          run: my-nginx
+      replicas: 2
```
What happens if you expose two pods on the same host on the same hostPort?

borkmann replied:
The second deployment remains in `Pending`: 
```
Events:
  Type     Reason            Age        From               Message
  ----     ------            ----       ----               -------
  Warning  FailedScheduling  <unknown>  default-scheduler  0/1 nodes are available: 1 node(s) didn't have free ports for the requested pod ports.
```
And once the first deployment is deleted, then it can proceed and also create the new HostPort service along with that.

brb replied:
Maybe then it makes sense to decrease the `replicas` count to 1, as the `Pending` state might confuse users who want to validate whether the feature works.

brb commented on install/kubernetes/cilium/values.yaml:
```diff
@@ -263,6 +263,11 @@ global:
+  # hostPort is the configuration for container hostPort mapping
+  hostPort:
+    # enabled enables the hostPort functionality
+    enabled: false
```
Do we want to disable the hostPort feature for new users? Asking, as for the existing users after the upgrade to v1.8 the feature will be enabled due to the agent defaulting the flag to `true`.

borkmann replied:
I would keep it enabled otherwise people might miss it and still deploy the `portmap` CNI chaining. I'll do a follow-up for the upgrade guide to document the support and actions needed.

brb reviewed (approved):
Nice! Couple nits and questions.
